cep18/src/utils.rs

```diff
@@ -0,0 +1,138 @@
+//! Implementation details.
+use core::convert::TryInto;
+
+use alloc::vec::Vec;
+use casper_contract::{
+    contract_api::{self, runtime, storage},
+    ext_ffi,
+    unwrap_or_revert::UnwrapOrRevert,
+};
+use casper_types::{
+    api_error,
+    bytesrepr::{self, FromBytes},
+    system::CallStackElement,
+    ApiError, CLTyped, Key, URef, U256,
+};
+
+use crate::{constants::TOTAL_SUPPLY, error::Cep18Error};
+
+/// Gets [`URef`] under a name.
+pub(crate) fn get_uref(name: &str) -> URef {
+    let key = runtime::get_key(name)
+        .ok_or(ApiError::MissingKey)
+        .unwrap_or_revert();
+    key.try_into().unwrap_or_revert()
+}
+
+/// Reads value from a named key.
+pub(crate) fn read_from<T>(name: &str) -> T
+where
+    T: FromBytes + CLTyped,
+{
+    let uref = get_uref(name);
+    let value: T = storage::read(uref).unwrap_or_revert().unwrap_or_revert();
+    value
+}
+
+/// Returns address based on a [`CallStackElement`].
+///
+/// For `Session` and `StoredSession` variants it will return account hash, and for `StoredContract`
+/// case it will use contract hash as the address.
+fn call_stack_element_to_address(call_stack_element: CallStackElement) -> Key {
+    match call_stack_element {
+        CallStackElement::Session { account_hash } => Key::from(account_hash),
+        CallStackElement::StoredSession { account_hash, .. } => {
+            // Stored session code acts in account's context, so if stored session wants to interact
+            // with an CEP18 token caller's address will be used.
+            Key::from(account_hash)
+        }
+        CallStackElement::StoredContract {
+            contract_package_hash,
+            ..
+        } => Key::from(contract_package_hash),
+    }
+}
+
+/// Gets the immediate session caller of the current execution.
+///
+/// This function ensures that only session code can execute this function, and disallows stored
+/// session/stored contracts.
```

Comment:
I know you didn't write this, but a bit confused here,
so the stored contracts couldn't call `transfer` or sth else?

Reply:
Why are we using the package hash for the caller here

Reply:
@gyroflaw not sure over time the impl and the doc might have deviated from each other.
@darthsiroftardis Because that was the original impl.

Reply:
> Why are we using the package hash for the caller here

I think we should use package hash, otherwise the users/devs will get confused regarding the usage between exist CEP18 and the new version.
tbh since we support upgradable smart contract feature, we should use contract package hash for contract instance not the specific version of contract hash